python/cudf/cudf/_lib/orc.pyx

```diff
@@ -268,24 +286,62 @@ cdef class ORCWriter:
         """
         Prepare all the values required to build the
         chunked_orc_writer_options anb creates a writer"""
-        cdef unique_ptr[table_metadata_with_nullability] tbl_meta
-        tbl_meta = make_unique[table_metadata_with_nullability]()
+        cdef table_view tv
 
         # Set the table_metadata
-        tbl_meta.get().column_names = get_column_names(table, self.index)
+        num_index_cols_meta = 0
+        self.tbl_meta = make_unique[table_input_metadata](
+            table_view_from_table(table, ignore_index=True)
+        )
+        if self.index is not False:
+            if isinstance(table._index, cudf.core.multiindex.MultiIndex):
```

Comment:
In `write_orc` the corresponding block checks for any non-RangeIndex type. Is that discrepancy intentional or meaningful? Or do you always get either a MultiIndex or a RangeIndex?

Reply:
I have to defer to @devavret or @galipremsagar on this one. The Python changes in this PR are based on the equivalent Parquet PR https://github.com/rapidsai/cudf/pull/7461.

Reply:
This code block is a replacement of the logic of `utils.get_column_names` which assigns the names of the columns to metadata. We don't want to update the metadata for RangeIndex because metadata contains only information on libcudf column_views. We could get a single series materialized index too, which is captured by the second part of the branch.

Regarding writing RangeIndex, this is currently broken #7011 (and re-reported in #9216) because it might take some smarts to merge range indices across different calls to `write_table`. The RangeIndex only gets written in the pandas specific metadata and passes straight through libcudf untouched. If you have some time, I'd invite you to take on #7011